store.go

```diff
@@ -1,199 +1,207 @@
+		record.SetID(NewID())
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
-
-	err := c.Insert(doc)
-	if err == nil {
-		doc.SetIsNew(false)
+	cols := s.schema.Columns()
+	values, err := RecordValues(record, cols...)
+	if err != nil {
+		return err
 	}
 
-	return err
-}
-
-// Update update the given document in the collection, returns error if a new
-// document is given.
-func (s *Store) Update(doc DocumentBase) error {
-	if doc.IsNew() {
-		return ErrNewDocument
+	_, err = s.inserter.
+		Columns(cols...).
+		Values(values...).
+		Exec()
+	if err != nil {
+		return err
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
-
-	return c.UpdateId(doc.GetId(), doc)
+	record.setPersisted(false)
+	return nil
 }
 
-// Save insert or update the given document in the collection, a document with
-// id should be provided. Upsert is used (http://godoc.org/gopkg.in/mgo.v2#Collection.Upsert)
-func (s *Store) Save(doc DocumentBase) (updated bool, err error) {
-	id := doc.GetId()
-	if len(id) == 0 {
-		return false, ErrEmptyID
+// Update updates the given fields of a record in the table. All fields are
+// updated if no fields are provided. For an update to take place, the record is
+// required to have a non-empty ID and not to be a new record.
+// Returns the number of updated rows and an error, if any.
+func (s *Store) Update(record Record, cols ...string) (int64, error) {
+	if record.IsPersisted() {
+		return 0, ErrNewDocument
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
+	if record.GetID().IsEmpty() {
+		return 0, ErrEmptyID
+	}
 
-	inf, err := c.UpsertId(id, doc)
+	if len(cols) == 0 {
+		cols = s.schema.Columns()
+	}
+	values, err := RecordValues(record, cols...)
 	if err != nil {
-		return false, err
+		return 0, err
 	}
 
-	doc.SetIsNew(false)
-	return inf.Updated > 0, nil
-}
+	var clauses = make(map[string]interface{}, len(cols))
+	for i, col := range cols {
+		clauses[col] = values[i]
+	}
 
-// Delete remove the document from the collection
-func (s *Store) Delete(doc DocumentBase) error {
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
+	result, err := s.updater.
+		SetMap(clauses).
+		Where(squirrel.Eq{s.schema.Identifier(): record.GetID()}).
+		Exec()
+	if err != nil {
+		return 0, err
+	}
 
-	return c.RemoveId(doc.GetId())
+	return result.RowsAffected()
 }
 
-// Find executes the given query in the collection
-func (s *Store) Find(q Query) (*ResultSet, error) {
-	sess, c := s.getSessionAndCollection()
-	mq := c.Find(q.GetCriteria())
-
-	if !q.GetSort().IsEmpty() {
-		mq.Sort(q.GetSort().ToList()...)
-	}
-
-	if q.GetSkip() != 0 {
-		mq.Skip(q.GetSkip())
+// Save inserts or updates the given record in the table. It requires a record
+// with non empty ID.
+func (s *Store) Save(record Record) (updated bool, err error) {
+	if record.GetID().IsEmpty() {
+		return false, ErrEmptyID
 	}
 
-	if q.GetLimit() != 0 {
-		mq.Limit(q.GetLimit())
+	if record.IsPersisted() {
```

Comment:
isNew != isPersisted
then... it should be 
`if !record.IsPersisted() {return false, s.Insert(record)}`